#! /bin/sh /usr/share/dpatch/dpatch-run
## clean-install-amp.dpatch by Diego Iastrubni <diego.iastrubni@xorcom.com>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Change some stupid defaults in freepbx, like wrong paths, making useless files
## DP: on the webroot, etc.
## DP: 
## DP: By default Debian's version of this script will not install files,
## DP: but only take chaege of installing the needed updates. I also removed
## DP: some dependencies checking, as they are already taken care by 
## DP: dpkg.

@DPATCH@
diff -urNad freepbx-2.2.1~dfsg~/install_amp freepbx-2.2.1~dfsg/install_amp
--- freepbx-2.2.1~dfsg~/install_amp	2007-02-07 02:40:46.000000000 +0200
+++ freepbx-2.2.1~dfsg/install_amp	2007-03-08 17:05:20.000000000 +0200
@@ -94,7 +94,7 @@
 	out("  --debug                  Enable debug output");
 	out("  --dry-run                Don't actually do anything");
 	out("  --force-version <ver>    Force upgrade from version <ver>");
-	out("  --no-files               Just run updates without installing files");
+	out("  --no-files               Just run updates without installing files (default)");
 	out("  --install-moh            Install default music-on-hold files (normally doesn't, unless ");
 	out("                           it's a new installation)");
 	out("  --my-svn-is-correct      Ignore Asterisk version, assume it is correct");
@@ -242,7 +242,7 @@
 	return ($retVal != 0);
 }
 
-function amp_mkdir($directory, $mode = "0755", $recursive = false) {
+function amp_mkdir($directory, $mode = "0775", $recursive = false) {
 	debug("mkdir ".$directory.", ".$mode);
 	$ntmp = sscanf($mode,"%o",$modenum); //assumes all inputs are octal
 	if (version_compare(phpversion(), 5.0) < 0) {
@@ -251,6 +251,7 @@
 			$output = false;
 			$return_value = false;
 			exec("mkdir -m ".$mode." -p ".$directory,  $output, $return_value);
+			exec("chown asterisk.asterisk ".$directory,  $output, $return_value);
 			return ($return_value == 0);
 		} else {
 			return mkdir($directory, $modenum);
@@ -312,7 +313,7 @@
 			if (is_dir($source)) {
 				if (!file_exists($destination)) {
 					if ((!$dryrun) && ($destination != "")) {
-						amp_mkdir($destination, "0750", true);
+						amp_mkdir($destination, "0775", true);
 					}
 				}
 			}
@@ -529,13 +530,13 @@
 	else $amp_conf["AMPMGRPASS"] = $key;
 	
 	do {
-		out("Enter the path to use for your AMP web root:\n [/var/www/html] ");
+		out("Enter the path to use for your AMP web root:\n [/var/www/] ");
 		$key = trim(fgets(STDIN,1024));
-		if (preg_match('/^$/',$key)) $amp_conf["AMPWEBROOT"] = "/var/www/html";
+		if (preg_match('/^$/',$key)) $amp_conf["AMPWEBROOT"] = "/var/www/";
 		else $amp_conf["AMPWEBROOT"] = rtrim($key,'/');
 		if (is_dir($amp_conf["AMPWEBROOT"])) {
 			break;
-		} else if (amp_mkdir($amp_conf["AMPWEBROOT"],"0755",true)){
+		} else if (amp_mkdir($amp_conf["AMPWEBROOT"],"0775",true)){
 			out("Created ".$amp_conf["AMPWEBROOT"]);
 			break;
 		} else {
@@ -562,13 +563,13 @@
 	else $amp_conf["AMPEXTENSIONS"] = $key;
 	
 	do {
-		out("Enter directory in which to store AMP executable scripts:\n [/var/lib/asterisk/bin] ");
+		out("Enter directory in which to store AMP executable scripts:\n [/usr/share/asterisk/bin] ");
 		$key = trim(fgets(STDIN,1024));
-		if (preg_match('/^$/',$key)) $amp_conf["AMPBIN"] = "/var/lib/asterisk/bin";
+		if (preg_match('/^$/',$key)) $amp_conf["AMPBIN"] = "/usr/share/asterisk/bin";
 		else $amp_conf["AMPBIN"] = rtrim($key,'/');
 		if (is_dir($amp_conf["AMPBIN"])) {
 			break;
-		} else if (amp_mkdir($amp_conf["AMPBIN"],"0755",true)){
+		} else if (amp_mkdir($amp_conf["AMPBIN"],"0775",true)){
 			out("Created ".$amp_conf["AMPBIN"]);
 			break;
 		} else {
@@ -583,7 +584,7 @@
 		else $amp_conf["AMPSBIN"] = rtrim($key,'/');
 		if (is_dir($amp_conf["AMPSBIN"])) {
 			break;
-		} else if (amp_mkdir($amp_conf["AMPSBIN"],"0755",true)){
+		} else if (amp_mkdir($amp_conf["AMPSBIN"],"0775",true)){
 			out("Created ".$amp_conf["AMPSBIN"]);
 			break;
 		} else {
@@ -613,7 +614,7 @@
 outn("Checking for PEAR DB..");
 if (! @ include('DB.php')) {
 	out("FAILED");
-	fatal("PEAR must be installed (requires DB.php). Include path: ".ini_get("include_path"));
+	fatal("PEAR must be installed (requires DB.php). Include path: ".ini_get("include_path") . "\nPlease install the package php4-pear or similar." );
 }
 out("OK");
 
@@ -622,7 +623,7 @@
 outn("Checking for PEAR Console::Getopt..");
 if (! @ include("Console/Getopt.php")) {
 	out("FAILED");
-	fatal("PEAR must be installed (requires Console/Getopt.php). Include path: ".ini_get("include_path"));
+	fatal("PEAR must be installed (requires Console/Getopt.php). Include path: ".ini_get("include_path") . "\nPlese install the package php4-pear or similar." );
 }
 out("OK");
 
@@ -639,7 +640,7 @@
 
 $debug = false;
 $dryrun = false;
-$install_files = true;
+$install_files = false;
 $override_astvers = false;
 
 $install_moh = false;
@@ -688,6 +689,10 @@
 			$install_files = false;
 			out("Running upgrade only, without installing files.");
 		break;
+		case "--install-files":
+			$install_files = true;
+			out("Installing files, I hope you know what you are doying....");
+		break;
 		case "--my-svn-is-correct":
 			$override_astvers = true;
 		break;
@@ -756,7 +761,7 @@
 if (!file_exists(AMP_CONF)) {
 	out(AMP_CONF." does not exist, copying default");
 	copy("amportal.conf", AMP_CONF);
-	chmod(AMP_CONF, 0640); 
+	chmod(AMP_CONF, 0660);
 	collect_settings(AMP_CONF, $dbhost, $new_username, $new_password, 'asterisk');
 
 	out("Assuming new install, --install-moh added to command line");
@@ -777,7 +782,7 @@
 
 if (!array_key_exists("AMPWEBROOT",$amp_conf)) {
 	out("Adding AMPWEBROOT option to amportal.conf - using AMP default");
-	$amp_conf["AMPWEBROOT"] = "/var/www/html";
+	$amp_conf["AMPWEBROOT"] = "/var/www/";
 }
 
 if (!array_key_exists("FOPWEBROOT",$amp_conf)) {
@@ -787,7 +792,7 @@
 
 if (!array_key_exists("AMPBIN",$amp_conf)) {
 	out("Adding AMPBIN option to amportal.conf - using AMP default");
-	$amp_conf["AMPBIN"] = "/var/lib/asterisk/bin";
+	$amp_conf["AMPBIN"] = "/usr/share/asterisk/bin";
 }
 
 if (!array_key_exists("AMPSBIN",$amp_conf)) {
@@ -823,7 +828,7 @@
 // write amportal.conf
 write_amportal_conf(AMP_CONF, $amp_conf);
 
-// **** Check for amportal.conf, create if necessary
+// **** Check for asterisk.conf, create if necessary
 
 outn("Checking for ".ASTERISK_CONF."..");
 if (!file_exists(ASTERISK_CONF)) {
@@ -846,7 +851,8 @@
    
    this code will stay in 2.2, but in 2.3 it will be gone. developers - please
    update your code
- */
+
+-- on debian this is not used at all.--
 if (isset($asterisk_conf['astetcdir'])) { $amp_conf['ASTETCDIR'] = $asterisk_conf['astetcdir']; }
 if (isset($asterisk_conf['astmoddir'])) { $amp_conf['ASTMODDIR'] = $asterisk_conf['astmoddir']; }
 if (isset($asterisk_conf['astvarlibdir'])) { $amp_conf['ASTVARLIBDIR'] = $asterisk_conf['astvarlibdir']; }
@@ -854,6 +860,7 @@
 if (isset($asterisk_conf['astspooldir'])) { $amp_conf['ASTSPOOLDIR'] = $asterisk_conf['astspooldir']; }
 if (isset($asterisk_conf['astrundir'])) { $amp_conf['ASTRUNDIR'] = $asterisk_conf['astrundir']; }
 if (isset($asterisk_conf['astlogdir'])) { $amp_conf['ASTLOGDIR'] = $asterisk_conf['astlogdir']; }
+ */
 
 if (!isset($pbx_engine)) { $pbx_engine='asterisk'; }
 out("Using $pbx_engine as PBX Engine");
@@ -869,13 +876,16 @@
 if ($exitcode != 0) {
 	fatal("Error executing asterisk: be sure Asterisk is installed and in the path");
 }
+/*
+-- on debian we do not do this thing --
+  see http://www.freepbx.org/trac/ticket/1679
 if (!$fd = fopen($amp_conf['ASTETCDIR'].'/version','w')) {
 	fatal('Cannot open '.$amp_conf['ASTETCDIR'].'/version for writing');
 }
 fwrite($fd, $tmpout);
 fclose($fd);
 // change to read-only
-chmod($amp_conf['ASTETCDIR'].'/version',0444);
+chmod($amp_conf['ASTETCDIR'].'/version',0444);*/
 
 
 // normally this would be the contents of ASTETCDIR/version, but this is for simplicity, as we just read it above
@@ -944,7 +954,7 @@
 		// datasource in in this style: dbengine://username:password@host/database 
 		if (!function_exists($db_engine.'_connect')) {
 			out("FAILED");
-			fatal($db_engine." PHP libraries not installed");
+			fatal($db_engine." PHP libraries not installed. \n Please install php4-mysql or similar.");
 		}
 	
 		$datasource = $db_engine.'://'.$db_user.':'.$db_pass.'@'.$db_host.'/'.$db_name;
@@ -1010,37 +1020,23 @@
 	$md5sums = read_md5_file(UPGRADE_DIR."/".$version.".md5");
 	recursive_copy("amp_conf", "", $md5sums);
 	if (!is_file("/etc/asterisk/voicemail.conf")) copy("/etc/asterisk/voicemail.conf.template","/etc/asterisk/voicemail.conf");
-	if (!is_dir("/var/spool/asterisk/voicemail/device")) amp_mkdir("/var/spool/asterisk/voicemail/device", "0755", true);
+	if (!is_dir("/var/spool/asterisk/voicemail/device")) amp_mkdir("/var/spool/asterisk/voicemail/device", "0775", true);
 	out("OK");
 }
 
 // **** Apply amportal.conf configuration to files
-debug("Running ".dirname(__FILE__)."/apply_conf.sh");
+debug("Running /usr/sbin/update-freepbx");
 outn("Configuring install for your environment..");
 if (!$dryrun) {
-	if (file_exists($amp_conf["AMPSBIN"]."/amportal"))
-		exec("chmod u+x ".$amp_conf["AMPSBIN"]."/amportal");
-	exec(dirname(__FILE__)."/apply_conf.sh");
+	exec("/usr/sbin/update-freepbx");
 }
 out("OK");
 
 // **** Create spool directories for monitor and fax
 if (!is_dir($asterisk_conf["astspooldir"]."/monitor"))
-	amp_mkdir($asterisk_conf["astspooldir"]."/monitor","0766",true);
+	amp_mkdir($asterisk_conf["astspooldir"]."/monitor","0775",true);
 if (!is_dir($asterisk_conf["astspooldir"]."/fax"))
-	amp_mkdir($asterisk_conf["astspooldir"]."/fax","0766",true);
-
-
-// **** Set permissions all files
-
-if ($install_files)
-{
-	outn("Setting permissions on files..");
-	if (!$dryrun) {
-		exec($amp_conf["AMPSBIN"]."/amportal chown");
-	}
-	out("OK");
-}
+	amp_mkdir($asterisk_conf["astspooldir"]."/fax","0775",true);
 
 
 // **** Read upgrades/ directory
@@ -1078,19 +1074,24 @@
 out("Generating AMP configs..OK");
 
 // **** Bounce FOP
+/*
+TODO still unavailable in debian
 outn("Restarting Flash Operator Panel..");
 exec('su - asterisk -c "'.$amp_conf["AMPWEBROOT"].'/admin/bounce_op.sh"');
 out("OK");
+*/
 
 $version = install_getversion();
+/*
+-- on debian we do not do this thing --
+  see http://www.freepbx.org/trac/ticket/1679
 $filename = $amp_conf["AMPWEBROOT"]."/admin/version.txt";
 if (!$fd = fopen($filename, "w")) {
-	fatal("Could not open ".$filename." for writing");
+       fatal("Could not open ".$filename." for writing");
 }
 fwrite($fd, $version);
 fclose($fd);
-
-
+*/
 
 // **** Set reload flag for AMP admin
 install_needreload();
